File: TeamCode/src/main/java/expo/util/Vector.java
package expo.util;

import java.util.Arrays;

public class Vector {
    private double x;
    private double y;
    private double theta;
    private double R;

    public Vector(double x, double y) {
        this.x = x;
        this.y = y;

        findTheta();
    }

    public void add(Vector other) {
        x += other.getX();
        y += other.getY();

        findTheta();
    }

    public void sub(Vector other) {
        x -= other.getX();
        y -= other.getY();

        findTheta();
    }


    public void rotate(double radians) {
        theta = (theta + radians) % (2 * Math.PI);
        findXY();
    }

    public void findXY() {
        x = R * Math.cos(theta);
        y = R * Math.sin(theta);
    }

    public void findTheta() {
        findXY();

        if (x == 0 && y == 0) {
            theta = 0;
        } else if (x == 0) {
            theta = Math.PI / 2;
            if (y < 0)
                theta *= -1;
        } else if (y == 0) {
            theta = Math.PI;
        } else {
            theta = Math.atan(y / x);
            if (y > 0) {
                theta += Math.PI;
            } else {
                theta -= Math.PI;
            }
        }
    }

    public void findR() {
        R = Math.sqrt(Math.pow(x, 2) + Math.pow(y, 2));
    }

    public double getX() {
        return x;
    }

    public double getY() {
        return y;
    }
}

File: TeamCode/src/main/java/expo/util/Vector.kt
package expo.util

import kotlin.math.*


class Vector(private var x: Double, private var y: Double) {
    private var theta = 0.0
    private var R = 0.0

    init {
        findTheta()
    }

    fun add(other: Vector) {
        x += other.getX()
        y += other.getY()
        findTheta()
    }

    fun sub(other: Vector) {
        x -= other.getX()
        y -= other.getY()
        findTheta()
    }


    fun rotate(radians: Double) {
        theta = (theta + radians) % (2 * Math.PI)
        findXY()
    }

    private fun findXY() {
        x = R * cos(theta)
        y = R * sin(theta)
    }

    private fun findTheta() {
        findR()
        if (x == 0.0 && y == 0.0) {
            theta = 0.0
        } else {
            if (x == 0.0) {
                theta = Math.PI / 2
                if (y < 0) {
                    theta *= -1.0
                }
            } else {
                theta = atan(y / x)
                if (x < 0) {
                    if (y >= 0) {
                        theta += Math.PI
                    } else if (y < 0) {
                        theta -= Math.PI
                    } else {
                        theta = Math.PI
                    }
                }
            }
        }
    }

    fun findR() {
        R = sqrt(x.pow(2) + y.pow(2))
    }

    fun getX(): Double {
        return x
    }

    fun getY(): Double {
        return y
    }

    override fun toString(): String {
        return "Vector(x=$x, y=$y)"
    }

    companion object {
        fun add(v1: Vector, v2: Vector): Vector {
            val copy = Vector(v1.getX(), v1.getY())
            copy.add(v2)
            return copy
        }

        fun sub(v1: Vector, v2: Vector): Vector {
            val copy = Vector(v1.getX(), v1.getY())
            copy.sub(v2)
            return copy
        }

    }
}